### PR TITLE
[Enhancement] [Refactor] (Multi Ref Base Table Part1) Enhance materialized view multi ref base table derivation

### PR DESCRIPTION
**Pull request by LiShuMing**
## Why I'm doing:
SR v3.3 starts to support multi ref base tables for asynchronous mv which can be partition change track for multi base tables. This is useful when mv's query contains  join or union all  operators.

But current multi ref base table derivation is limited:
- don't support nested union all (only support union all's children with scan or join)
- only support date_trunc partition expression.

## What I'm doing:
- Refactor MvJoinFilter.java  into `MVPartitionExprResolver`
- Enhance materialized view multi ref base table derivation to support more patterns. ege: Those mvs can use multi ref base tables optimaztion which can not be used before:
```
CREATE MATERIALIZED VIEW IF NOT EXISTS `test_mv3`
PARTITION BY date_trunc('day', dt) 
DISTRIBUTED BY HASH(`dt`)
REFRESH ASYNC 
AS 
select dt, sum(s_id) as s_id
from
(
    select dt, sum(id) as s_id from u1 group by dt
    union all
    select dt, sum(id) as s_id from u2 group by dt
) t group by dt;

```
- Depredated  original single ref  table method `getRefBaseTablePartitionColumn`, use `getRefBaseTablePartitionColumnMap` as default. 

Fixes https://github.com/StarRocks/StarRocksTest/issues/8101

https://github.com/StarRocks/starrocks/issues/48354

### Further
- This PR will introduce a problem with transparent mv which has not supported multi ref base tables. I will fix it later.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/analysis/Exprs.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License
+package com.starrocks.analysis;
+
+import com.google.api.client.util.Lists;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+public class Exprs {
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/analysis/Exprs.java:21]: it seems that Exprs has nothing different from List<Expr>, why need introduce this class?

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 2]
```diff
@@ -466,9 +470,16 @@ public String toString() {
     @SerializedName(value = "maxMVRewriteStaleness")
     private int maxMVRewriteStaleness = 0;
 
+    // Multi ref base table partition expression and ref slot ref map.
     @SerializedName(value = "partitionExprMaps")
     private Map<ExpressionSerializedObject, ExpressionSerializedObject> serializedPartitionExprMaps;
     private Map<Expr, SlotRef> partitionExprMaps;
+    // ref base table to partition expression
+    private Optional<Map<Table, Expr>> refBaseTablePartitionExprsOpt = Optional.empty();
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:478]: it seems that Optional should be used as type of parameter passing in/out functions. 
class field definition should not adopt Optional type.

use transient keyword before cached fields.

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 3]
```diff
@@ -1407,36 +1432,49 @@ public boolean containsBaseTable(TableName tableName) {
...
+                    LOG.info("Base table {} contains no partition expr, skip", table.getName());
+                    continue;
                 }
+                refBaseTablePartitionExprMap.put(table, mvPartitionExpr.get().getExpr());
             }
+            LOG.info("The refBaseTablePartitionExprMap of mv {} is {}", getName(), refBaseTablePartitionExprMap);
+            refBaseTablePartitionExprsOpt = Optional.of(refBaseTablePartitionExprMap);
         }
-        return tableToJoinExprMap;
+        return refBaseTablePartitionExprsOpt.get();
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:1454]: add Precondition.check(refBaseTablePartitionExprsOpt.isPresent());

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 4]
```diff
@@ -1520,72 +1558,83 @@ private boolean isManyToManyPartitionRangeMapping(PRangeCell srcRange,
     /**
      * Materialized View's partition column can only refer one base table's partition column, get the referred
      * base table and its partition column.
-     * TODO: support multi-column partitions later.
-     *
      * @return : The materialized view's referred base table and its partition column.
+     * TODO: Support multi-column partitions later.
+     * TODO: Use getRefBaseTablePartitionColumnMap instead since SR v3.3 has supported multi-tables partition change tracking.
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:1563]: getRefBaseTablePartitionColumns?

**Review by satanson — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by stephen-shelby — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Pull request merged**